containers/src/Data/IntSet/Internal.hs

```diff
@@ -1486,8 +1487,94 @@ equal _   _   = False
 --------------------------------------------------------------------}
 
 instance Ord IntSet where
-    compare s1 s2 = compare (toAscList s1) (toAscList s2)
-    -- tentative implementation. See if more efficient exists.
+  compare = compareIntSets
+
+compareIntSets :: IntSet -> IntSet -> Ordering
+compareIntSets s1 s2 = case (splitSign s1, splitSign s2) of
+  ((l1, r1), (l2, r2)) -> case go l1 l2 of
+    A_LT_B -> LT
+    A_Prefix_B -> if null r1 then LT else GT
+    A_EQ_B -> case go r1 r2 of
+      A_LT_B -> LT
+      A_Prefix_B -> LT
+      A_EQ_B -> EQ
+      B_Prefix_A -> GT
+      A_GT_B -> GT
+    B_Prefix_A -> if null r2 then GT else LT
+    A_GT_B -> GT
+  where
+    go t1@(Bin p1 l1 r1) t2@(Bin p2 l2 r2) = case treeTreeBranch p1 p2 of
+      ABL -> case go l1 t2 of
+        A_Prefix_B -> A_GT_B
+        A_EQ_B -> B_Prefix_A
+        o -> o
+      ABR -> A_LT_B
+      BAL -> case go t1 l2 of
+        A_EQ_B -> A_Prefix_B
+        B_Prefix_A -> A_LT_B
+        o -> o
+      BAR -> A_GT_B
+      EQL -> case go l1 l2 of
+        A_Prefix_B -> A_GT_B
+        A_EQ_B -> go r1 r2
+        B_Prefix_A -> A_LT_B
+        o -> o
+      NOM -> if unPrefix p1 < unPrefix p2 then A_LT_B else A_GT_B
+    go (Bin _ l1 _) (Tip k2 bm2) = case leftmostTipSure l1 of
+      Tip' k1 bm1 -> case orderTips k1 bm1 k2 bm2 of
+        A_Prefix_B -> A_GT_B
+        A_EQ_B -> B_Prefix_A
+        o -> o
+    go (Tip k1 bm1) (Bin _ l2 _) = case leftmostTipSure l2 of
+      Tip' k2 bm2 -> case orderTips k1 bm1 k2 bm2 of
+        A_EQ_B -> A_Prefix_B
+        B_Prefix_A -> A_LT_B
+        o -> o
+    go (Tip k1 bm1) (Tip k2 bm2) = orderTips k1 bm1 k2 bm2
+    go Nil Nil = A_EQ_B
+    go Nil _ = A_Prefix_B
+    go _ Nil = B_Prefix_A
+
+-- This type allows GHC to return unboxed ints from leftmostTipSure, as
+-- $wleftmostTipSure :: IntSet -> (# Int#, Word# #)
+-- On a modern enough GHC (>=9.4) this is unnecessary, we could use StrictPair
+-- instead and get the same Core.
+data Tip' = Tip' {-# UNPACK #-} !Int {-# UNPACK #-} !BitMap
+
+leftmostTipSure :: IntSet -> Tip'
```

Comment:
Please add a Haddock string.

Reply:
Why? These functions are not exposed.

Reply:
@treeowl do you still want to add doc strings?

Reply:
I'll try to write one for `compareTips` by tomorrow night. But otherwise I guess you can merge and I'll open an issue to remember.

Reply:
I'll go ahead and merge it then.